crates/feldera-storage/src/backend/glommio_impl/mod.rs

```diff
@@ -89,17 +117,13 @@ impl StorageWrite for GlommioBackend {
     }
 
     async fn complete(&self, fd: FileHandle) -> Result<ImmutableFileHandle, StorageError> {
-        let mut files = self.files.write().await?;
-
-        let file = files.remove(&fd.0).unwrap();
-        file.fdatasync().await?;
-        let mut path_buf = PathBuf::new();
-        file.path().unwrap().clone_into(&mut path_buf);
-        file.close().await?;
-
-        let readable_file = DmaFile::open(path_buf).await?;
-        files.insert(fd.0, readable_file);
-
+        self.files
```

Comment:
This means the file remains in `self.files` even if fdatasync fails. But probably not an issue aside from having some garbage in `files` since we loose the `fd` and never get an `ImmutableFileHandle` to read it.

Reply:
Hmm, I hadn't considered that.

Reply:
I'll force-push an updated branch in a minute that folds this in to avoid the leak:

```patch
diff --git a/crates/feldera-storage/src/backend/glommio_impl/mod.rs b/crates/feldera-storage/src/backend/glommio_impl/mod.rs
index ece7ba4e..3e8ec0ad 100644
--- a/crates/feldera-storage/src/backend/glommio_impl/mod.rs
+++ b/crates/feldera-storage/src/backend/glommio_impl/mod.rs
@@ -117,13 +117,12 @@ impl StorageWrite for GlommioBackend {
     }
 
     async fn complete(&self, fd: FileHandle) -> Result<ImmutableFileHandle, StorageError> {
-        self.files
-            .read()
-            .await?
-            .get(&fd.0)
-            .unwrap()
-            .fdatasync()
-            .await?;
+        let mut files = self.files.write().await?;
+
+        let file = files.remove(&fd.0).unwrap();
+        file.fdatasync().await?;
+        files.insert(fd.0, file);
+
         Ok(ImmutableFileHandle(fd.0))
     }
 }
diff --git a/crates/feldera-storage/src/backend/monoio_impl/mod.rs b/crates/feldera-storage/src/backend/monoio_impl/mod.rs
index 0f0a27f0..e5b61495 100644
--- a/crates/feldera-storage/src/backend/monoio_impl/mod.rs
+++ b/crates/feldera-storage/src/backend/monoio_impl/mod.rs
@@ -154,14 +154,12 @@ impl StorageWrite for MonoioBackend {
     }
 
     async fn complete(&self, fd: FileHandle) -> Result<ImmutableFileHandle, StorageError> {
-        self.files
-            .read()
-            .await
-            .get(&fd.0)
-            .unwrap()
-            .file
-            .sync_all()
-            .await?;
+        let mut files = self.files.write().await;
+
+        let fm = files.remove(&fd.0).unwrap();
+        fm.file.sync_all().await?;
+        files.insert(fd.0, fm);
+
         Ok(ImmutableFileHandle(fd.0))
     }
 }
```